src/main/__test__/context-handler.test.ts

```diff
@@ -0,0 +1,114 @@
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import { UserStore } from "../../common/user-store";
+import { ContextHandler } from "../context-handler";
+import { PrometheusProvider, PrometheusProviderRegistry, PrometheusService } from "../prometheus";
+
+enum AlwaysFail {
+  No,
+  Throw,
+  Return,
+}
+
+class TestProvider extends PrometheusProvider {
+  name = "TestProvider1";
+  rateAccuracy = "1h";
+  isConfigurable = false;
+
+  constructor(public id: string, public alwaysFail = AlwaysFail.No) {
+    super();
+  }
+
+  getQuery(): string {
+    throw new Error("getQuery is not implemented.");
+  }
+
+  async getPrometheusService(): Promise<PrometheusService> {
+    switch (this.alwaysFail) {
+      case AlwaysFail.No:
+        return {
+          id: this.id,
+          namespace: "default",
+          port: 7000,
+          service: "",
+        };
+      case AlwaysFail.Throw:
+        throw new Error("does fail");
+      case AlwaysFail.Return:
+        return undefined;
+    }
+  }
+}
+
+function getHandler() {
+  return new ContextHandler(({
+    getProxyKubeconfig: (): any => ({
+      makeApiClient: (): any => undefined,
+    }),
+    apiUrl: "http://localhost:81",
+  }) as any);
+}
+
+describe("ContextHandler", () => {
+  beforeEach(() => {
+    PrometheusProviderRegistry.createInstance();
+    UserStore.createInstance();
+  });
+
+  afterEach(() => {
+    PrometheusProviderRegistry.resetInstance();
+    UserStore.resetInstance();
+  });
+
+  describe("getPrometheusService", () => {
+    it("should pick the first provider of 1 success", async () => {
+      PrometheusProviderRegistry.getInstance()
+        .registerProvider(new TestProvider("id_foobar"));
+
+      const service = await getHandler().getPrometheusService();
+
+      expect(service.id === "id_foobar");
+    });
+
+    it("should pick the first provider of 1 success, after 2 fails", async () => {
+      PrometheusProviderRegistry.getInstance()
+        .registerProvider(new TestProvider("id_foobar_1", AlwaysFail.Throw))
+        .registerProvider(new TestProvider("id_foobar_2", AlwaysFail.No))
+        .registerProvider(new TestProvider("id_foobar_3"));
+
+      const service = await getHandler().getPrometheusService();
+
+      expect(service.id === "id_foobar_3");
+    });
+
+    it("should pick the first provider of 1 success, between 2 fails", async () => {
+      PrometheusProviderRegistry.getInstance()
+        .registerProvider(new TestProvider("id_foobar_1", AlwaysFail.Throw))
+        .registerProvider(new TestProvider("id_foobar_2"))
+        .registerProvider(new TestProvider("id_foobar_3", AlwaysFail.No));
+
+      const service = await getHandler().getPrometheusService();
+
+      expect(service.id === "id_foobar_2");
```

Comment:
this should've failed then?

Reply:
No this is still correct because `AlwaysFail.No` is the default, and by design we pick the first one that succeeds.

Reply:
Maybe? will investigate.